tensorflow/compiler/xla/service/gpu/gpu_fused_mha_runner.h

```diff
@@ -0,0 +1,241 @@
+                                config.algorithm))
+                      : config.kind == CudnnfMHAKind::kScaleMaskSoftmax ||
+                                config.kind ==
+                                    CudnnfMHAKind::kScaleMaskSoftmaxDropout
+                            ? FusedMultiHeadedAttentionRunner(
+                                  std::make_unique<se::dnn::LazyOpRunner<
+                                      se::dnn::FusedMHAScaleMaskSoftmaxOp>>(
+                                      config.algorithm))
+                            : FusedMultiHeadedAttentionRunner(
+                                  std::make_unique<se::dnn::LazyOpRunner<
+                                      se::dnn::FusedMHAScaleBiasMaskSoftmaxOp>>(
+                                      config.algorithm))) {}
+
+  se::dnn::AlgorithmDesc ToAlgorithmDesc() const {
+    return std::visit(ToAlgorithmDescVisitor{}, repr_);
+  }
+
+  se::dnn::LazyOpRunner<se::dnn::FusedMHASoftmaxOp>* AsFusedMHASoftmaxRunner() {
+    CHECK(std::holds_alternative<
+          std::unique_ptr<se::dnn::LazyOpRunner<se::dnn::FusedMHASoftmaxOp>>>(
+        repr_));
+    return std::
+        get<std::unique_ptr<se::dnn::LazyOpRunner<se::dnn::FusedMHASoftmaxOp>>>(
+               repr_)
+            .get();
+  }
+
+  se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleMaskSoftmaxOp>*
+  AsFusedMHAMaskRunner() {
+    CHECK(std::holds_alternative<std::unique_ptr<
+              se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleMaskSoftmaxOp>>>(
+        repr_));
+    return std::get<std::unique_ptr<
+        se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleMaskSoftmaxOp>>>(repr_)
+        .get();
+  }
+
+  se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleBiasMaskSoftmaxOp>*
+  AsFusedMHABiasMaskRunner() {
+    CHECK(std::holds_alternative<std::unique_ptr<
+              se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleBiasMaskSoftmaxOp>>>(
+        repr_));
+    return std::get<std::unique_ptr<
+        se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleBiasMaskSoftmaxOp>>>(repr_)
+        .get();
+  }
+
+  se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleBiasSoftmaxOp>*
+  AsFusedMHABiasRunner() {
+    CHECK(std::holds_alternative<std::unique_ptr<
+              se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleBiasSoftmaxOp>>>(
+        repr_));
+    return std::get<std::unique_ptr<
+        se::dnn::LazyOpRunner<se::dnn::FusedMHAScaleBiasSoftmaxOp>>>(repr_)
+        .get();
+  }
+
+ private:
+  struct ToAlgorithmDescVisitor {
+    template <typename RunnerPtr>
+    se::dnn::AlgorithmDesc operator()(const RunnerPtr& runner) {
+      return runner->ToAlgorithmDesc();
+    }
+
+    se::dnn::AlgorithmDesc operator()(const std::monostate&) {
+      CHECK(false) << "Internal error: uninitialized runner in ToAlgorithmDesc";
+    }
+  };
+  using Repr = std::variant<
+      std::monostate,  // To allow XXX default ctor
+      std::unique_ptr<se::dnn::LazyOpRunner<se::dnn::FusedMHASoftmaxOp>>,
```

Comment:
Nit: std::unique_ptr<std::variant> would make more sense and also would automatically allow for empty state with no need for monostate.

Reply:
I agree but I've actually followed the same convention as convolution and other instances found in the codebase. Do you want to have them changed as well? Moreover this will need handing of the variable differently in quite a few places. 
If that's ok with you, I'd prefer to keep it this way for now and may be add a refactoring of all such instances (including conv) as a separate PR if needed.